File: src/main/java/com/zepben/vertxutils/routing/ChunkedResponse/ChunkedJsonResponse.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing.ChunkedResponse;

import com.zepben.annotations.EverythingIsNonnullByDefault;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Fluent Helper to send a JSON response in chunks.
 */
@SuppressWarnings("WeakerAccess")
@EverythingIsNonnullByDefault
public abstract class ChunkedJsonResponse {

    public final static int DEFAULT_BUFFER_SIZE = 1 << 21;

    public class JsonObject {

        private JsonObject() {
        }

        public JsonArray beginArray(String key) {
            if (isNotObject())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open object to begin an array.");

            return doAddKey(key).doBeginArray();
        }

        public JsonObject beginObject(String key) {
            if (isNotObject())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open object to begin another object.");

            return doAddKey(key).doBeginObject();
        }

        public JsonObject addJson(String key, String json) {
            if (isNotObject())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open object to add json.");

            return doAddKey(key).doAddJson(json);
        }

        public JsonObject endObject() {
            if (isNotObject())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open object to end.");

            doEndObject(JsonObjectType.OBJECT);
            return object;
        }

        public JsonArray endObjectInArray() {
            if (isNotObject())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open object to end.");

            doEndObject(JsonObjectType.ARRAY);
            return array;
        }

        @SuppressWarnings("UnusedReturnValue")
        public JsonObject send(boolean force) {
            ChunkedJsonResponse.this.send(force, sb);
            return this;
        }

        private JsonArray doBeginArray() {
            openItems.push(JsonObjectType.ARRAY);
            isFirst = true;

            sb.append("[");

            return array;
        }

        private JsonObject doBeginObject() {
            openItems.push(JsonObjectType.OBJECT);
            isFirst = true;

            sb.append("{");

            return object;
        }

        private JsonObject doAddKey(String key) {
            if (isFirst)
                isFirst = false;
            else
                sb.append(",");

            sb.append("\"").append(key).append("\":");

            return this;
        }

        private JsonObject doAddJson(String json) {
            sb.append(json);
            send(false);

            return this;
        }

        private void doEndObject(JsonObjectType expectedParentType) {
            openItems.pop();
            if (!openItems.isEmpty() && (openItems.peek() != expectedParentType)) {
                openItems.push(JsonObjectType.OBJECT);
                throw new IllegalStateException("INTERNAL ERROR: Incorrect end object method called, the parent is not of the expected type.");
            }

            sb.append("}");
            isFirst = false;

            if (openItems.isEmpty())
                end(sb);
            else
                send(false);
        }

    }

    public class JsonArray {

        private JsonArray() {
        }

        public JsonArray addArrayItem(String json) {
            if (isNotArray())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open array to add json.");

            if (isFirst)
                isFirst = false;
            else
                sb.append(",");

            sb.append(json);
            send(false);

            return this;
        }

        public JsonObject beginObject() {
            if (isNotArray())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open array to begin an object.");

            openItems.push(JsonObjectType.OBJECT);

            if (!isFirst)
                sb.append(",");

            isFirst = true;
            sb.append("{");

            return object;
        }

        public JsonArray beginArray() {
            if (isNotArray())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open array to begin an array.");

            openItems.push(JsonObjectType.ARRAY);

            if (!isFirst)
                sb.append(",");

            isFirst = true;
            sb.append("[");

            return array;
        }

        public JsonObject endArray() {
            if (isNotArray())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open array to end.");

            doEndArray(JsonObjectType.OBJECT);
            return object;
        }

        public JsonArray endArrayInArray() {
            if (isNotArray())
                throw new IllegalStateException("INTERNAL ERROR: You must have an open array to end.");

            doEndArray(JsonObjectType.ARRAY);
            return array;
        }

        @SuppressWarnings("UnusedReturnValue")
        public JsonArray send(boolean force) {
            ChunkedJsonResponse.this.send(force, sb);
            return this;
        }

        private void doEndArray(JsonObjectType expectedParentType) {
            openItems.pop();
            if (!openItems.isEmpty() && (openItems.peek() != expectedParentType)) {
                openItems.push(JsonObjectType.ARRAY);
                throw new IllegalStateException("INTERNAL ERROR: Incorrect end array method called, the parent is not of the expected type.");
            }

            sb.append("]");
            isFirst = false;

            if (openItems.isEmpty())
                end(sb);
            else
                send(false);
        }

    }

    private enum JsonObjectType {OBJECT, ARRAY}


    private final StringBuilder sb;
    private final JsonObject object = new JsonObject();
    private final JsonArray array = new JsonArray();

    boolean isFirst = true;
    private final Deque<JsonObjectType> openItems = new ArrayDeque<>();

    public ChunkedJsonResponse(int bufferSize) {
        sb = new StringBuilder(bufferSize);
    }

    public JsonArray ofArray() {
        if (!openItems.isEmpty())
            throw new IllegalStateException("INTERNAL ERROR: You can only start one object or array for a response.");
        return object.doBeginArray();
    }

    public JsonObject ofObject() {
        if (!openItems.isEmpty())
            throw new IllegalStateException("INTERNAL ERROR: You can only start one object or array for a response.");
        return object.doBeginObject();
    }

    private boolean isNotArray() {
        return openItems.peek() != JsonObjectType.ARRAY;
    }

    private boolean isNotObject() {
        return openItems.peek() != JsonObjectType.OBJECT;
    }

    protected abstract void end(StringBuilder sb);

    protected abstract void send(boolean force, StringBuilder sb);

}

File: src/main/java/com/zepben/vertxutils/routing/chunked/ChunkedJsonResponse.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing.chunked

import io.vertx.core.json.Json
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject

/**
 * The base class for building chunked JSON responses.
 *
 * @param bufferSize The initial capacity of the underlying string builder.
 */
abstract class ChunkedJsonResponse(
    bufferSize: Int = DEFAULT_BUFFER_SIZE,
) {

    companion object {

        const val DEFAULT_BUFFER_SIZE = 1 shl 21

    }

    private val sb = StringBuilder(bufferSize)
    private val needsSeparatorStack = ArrayDeque<Boolean>()

    /**
     * DSL entry point for building a JSON object response.
     *
     * @param block The DSL block used to populate the object contents.
     */
    fun ofObject(block: JsonObjectBuilder.() -> Unit) {
        check(sb.isEmpty()) { "Can't reuse a non-clean response builder" }
        JsonObjectBuilder().build(block)
    }

    /**
     * DSL entry point for building a JSON array response.
     *
     * @param block The DSL block used to populate the array contents.
     */
    fun ofArray(block: JsonArrayBuilder.() -> Unit) {
        check(sb.isEmpty()) { "Can't reuse a non-clean response builder" }
        JsonArrayBuilder().build(block)
    }

    /**
     * A DSL builder for a JSON object.
     */
    inner class JsonObjectBuilder : JsonBuilder<JsonObjectBuilder>("{", "}") {

        /**
         * Add a scalar field to the JSON object being built, with appropriate escaping.
         *
         * NOTE: This deliberately removes support for adding JsonObject and JsonArray values directly, they should be provided via DSL builders.
         *
         * @param key The key for the filed in the object.
         * @param value The value to associate with the [key].
         * @throws IllegalArgumentException for any unsupported value types.
         */
        fun field(key: String, value: Any?) {
            sb.writeKey(key).appendJsonValue(value)
            checkWrite(sb)
        }

        /**
         * Add a nested object to the JSON object being built.
         *
         * @param key The key for the nested object in the object.
         * @param block The DSL block used to build the nested object.
         */
        fun obj(key: String, block: JsonObjectBuilder.() -> Unit) {
            sb.writeKey(key)
            JsonObjectBuilder().build(block)
        }

        /**
         * Add a nested array to the JSON object being built.
         *
         * @param key The key for the nested array in the object.
         * @param block The DSL block used to build the nested array.
         */
        fun array(key: String, block: JsonArrayBuilder.() -> Unit) {
            sb.writeKey(key)
            JsonArrayBuilder().build(block)
        }

        private fun StringBuilder.writeKey(key: String): StringBuilder =
            maybeAppendSeparator()
                .append(Json.encode(key))
                .append(":")

    }

    /**
     * A DSL builder for a JSON array.
     */
    inner class JsonArrayBuilder : JsonBuilder<JsonArrayBuilder>("[", "]") {

        /**
         * Add a scalar item to the JSON array being built, with appropriate escaping.
         *
         * NOTE: This deliberately removes support for adding JsonObject and JsonArray values directly, they should be provided via DSL builders.
         *
         * @param value The value to add to the array.
         * @throws IllegalArgumentException for any unsupported value types.
         */
        fun item(value: Any?) {
            sb.maybeAppendSeparator().appendJsonValue(value)
            checkWrite(sb)
        }

        /**
         * Add a nested object to the JSON array being built.
         *
         * @param block The DSL block used to build the nested object.
         */
        fun obj(block: JsonObjectBuilder.() -> Unit) {
            sb.maybeAppendSeparator()
            JsonObjectBuilder().build(block)
        }

        /**
         * Add a nested array to the JSON array being built.
         *
         * @param block The DSL block used to build the nested array.
         */
        fun array(block: JsonArrayBuilder.() -> Unit) {
            sb.maybeAppendSeparator()
            JsonArrayBuilder().build(block)
        }

    }

    /**
     * A base class for the DSL builders.
     *
     * @param openToken The token used to open the element being built by this builder.
     * @param openToken The token used to close the element being built by this builder.
     */
    abstract inner class JsonBuilder<B : JsonBuilder<B>>(
        private val openToken: String,
        private val closeToken: String,
    ) {

        /**
         * Write the current buffer to the underlying destination if it is appropriate (e.g. exceeds minimum size requirements), or if it is forced.
         *
         * @param force Flag to indicate that writing should occur without performing any other checks.
         */
        fun checkWrite(force: Boolean = false) = checkWrite(sb, force)

        /**
         * Build the element with the given DSL block.
         *
         * @param block The DSL block used to populate this element.
         */
        internal fun build(block: B.() -> Unit) {
            sb.append(openToken)
            needsSeparatorStack.addLast(false)

            @Suppress("UNCHECKED_CAST")
            (this as B).block()

            sb.append(closeToken)
            needsSeparatorStack.removeLast()

            if (needsSeparatorStack.isEmpty())
                onResponseCompleted(sb)
            else
                checkWrite(sb)
        }

        /**
         * Append a separator if it is required before adding the next element.
         *
         * The first item added to any element will suppress the separator, with any subsequent elements inserting it.
         */
        protected fun StringBuilder.maybeAppendSeparator(): StringBuilder {
            if (needsSeparatorStack.isNotEmpty()) {
                if (needsSeparatorStack.removeLast())
                    append(",")
                needsSeparatorStack.addLast(true)
            }
            return this
        }

        /**
         * Append a JSON value to the element. This will provide required escaping.
         *
         * @param value The scalar value to add to this element.
         * @throws IllegalArgumentException for any unsupported value types.
         */
        protected fun StringBuilder.appendJsonValue(value: Any?): StringBuilder =
            when (value) {
                null -> append("null")
                is String -> append(Json.encode(value))
                is Number, is Boolean -> append(value.toString())
                is JsonObject -> append(value.encode())
                is JsonArray -> append(value.encode())
                else -> throw IllegalArgumentException("Unsupported JSON value type: ${value::class}")
            }

    }

    /**
     * Write the current buffer to the underlying destination if it is appropriate (e.g. exceeds minimum size requirements), or if it is forced.
     *
     * NOTE: The [sb] buffer won't be reset by the base class, so this should be done in this function if required for the implementation.
     *
     * @param sb The buffer to write if required.
     * @param force Flag to indicate that writing should occur without performing any other checks.
     */
    protected abstract fun checkWrite(sb: StringBuilder, force: Boolean = false)

    /**
     * Notification the response has been completed.
     *
     * This should handle the reaming buffer, then reset it if the response can be reused.
     *
     * @param sb The remaining buffer when the response was completed.
     */
    protected abstract fun onResponseCompleted(sb: StringBuilder)

}
